jobrunner/local_docker_job_executor.py

```diff
@@ -0,0 +1,451 @@
+# helpful for building tooling for inspecting live processes.
+JOB_LABEL = "jobrunner-job"
+
+
+def get_status_of_finished_job(job_id, workspace, action, output_spec):
+    try:
+        container_metadata = get_container_metadata(job_id)
+        outputs, unmatched_patterns = find_matching_outputs(job_id, output_spec)
+        # Set the final state of the job
+        status_code = None
+        unmatched_outputs = []
+        if container_metadata["State"]["ExitCode"] != 0:
+            state = State.FAILED
+            status_message = "Job exited with an error code"
+            status_code = StatusCode.NONZERO_EXIT
+        elif unmatched_patterns:
+            state = State.FAILED
+            status_message = "No outputs found matching patterns:\n - {}".format(
+                "\n - ".join(unmatched_patterns)
+            )
+            # If the job fails because an output was missing its very useful to
+            # show the user what files were created as often the issue is just a
+            # typo
+            unmatched_outputs = get_unmatched_outputs(job_id, outputs)
+        else:
+            state = State.SUCCEEDED
+            status_message = "Completed successfully"
+        # job_metadata is a big dict capturing everything we know about the state
+        # of the job
+        job_metadata = get_job_metadata(container_metadata)
+        # Dump useful info in log directory
+        log_dir = get_log_dir(job_id)
+        ensure_overwritable(log_dir / "logs.txt", log_dir / "metadata.json")
+        write_log_file(job_id, unmatched_outputs, job_metadata, log_dir / "logs.txt")
+        with open(log_dir / "metadata.json", "w") as f:
+            json.dump(job_metadata, f, indent=2)
+        # Copy logs to workspace
+        workspace_dir = get_high_privacy_workspace(workspace)
+        metadata_log_file = workspace_dir / METADATA_DIR / f"{action}.log"
+        copy_file(log_dir / "logs.txt", metadata_log_file)
+        log.info(f"Logs written to: {metadata_log_file}")
+        # Extract outputs to workspace
+        ensure_overwritable(*[workspace_dir / f for f in outputs.keys()])
+        volume = volume_name(job_id)
+        for filename in outputs.keys():
+            log.info(f"Extracting output file: {filename}")
+            docker.copy_from_volume(volume, filename, workspace_dir / filename)
+        # Copy out logs and medium privacy files
+        medium_privacy_dir = get_medium_privacy_workspace(workspace)
+        if medium_privacy_dir:
+            copy_file(
+                workspace_dir / METADATA_DIR / f"{action}.log",
+                medium_privacy_dir / METADATA_DIR / f"{action}.log",
+            )
+            for filename, privacy_level in outputs.items():
+                if privacy_level == "moderately_sensitive":
+                    copy_file(workspace_dir / filename, medium_privacy_dir / filename)
+        cleanup_job(job_id)
```

Comment:
This makes the function non-idempotent. We can't safely clean up until the caller has persisted its state, in whatever form it does that.

Unless I'm missing something, the only safe way to do this is to have an additional API call with the semantics of "make job garbage-collectable". Whether that's done via an inline operation, or just updating some metadata on the job or whatever it may be, is up the executor. But it needs to be the job-runner that tells it that it's OK to do so.

Reply:
Ooh interesting. Thanks. I'll have a think about it.

Reply:
Ugh. Distributed systems are hard.